Improve contribution guide (#6140)
chainer/chainer: docs/source/contribution.rst

   The issue tracker is not a place to share knowledge on practices.

How to Send a Pull Request
~~~~~~~~~~~~~~~~~~~~~~~~~~

If you can write code to fix an issue, it is encouraged to send a PR.

In that case, confirm the following points before starting to write any code.

- Read the :ref:`coding-guide` and the :ref:`testing-guide`.
- Check the appropriate branch to which you should send a PR, following :ref:`contrib-git-branches`.
  If you are unsure about which branch to target, choose the ``master`` branch.
  The current source tree of the chosen branch is the starting point of your change.

After writing your code **(including unit tests and hopefully documentations!)**, send a PR on GitHub.
You have to write a precise explanation of **what** and **how** in the description;
this is the first documentation of your code and an important part of your PR.

However, even if your code is not complete, you can send a PR as a *work-in-progress (WIP) PR* by prefixing the PR title with ``[WIP]``.
If you just describe the PR, the core team and other contributors can join the discussion about how to proceed with it.

> WIP PRs may occasionally be be useful for discussing based on concrete code.

Reviewer: ```suggestion
WIP PRs may occasionally be useful for discussing based on concrete code.
```

> When a PR is created (or updated), it is automatically tested in the CI environment.

Reviewer: Is it about Travis CI? If so, the next line is inaccurate.

Author: Clarified it a bit. Can you take a look?

Reviewer: Don't we need to mention triggering of other CIs? (just writing "manually by a committer" might be ok)

Author: Updated again, PTAL.

Once tests pass and the PR is approved by a reviewer from the core team, it will be merged.

.. note::

    If you are planning to add a new feature or modify existing APIs, **it is recommended to open an issue and discuss the design first.**
    Following the consequences of the discussions, you can send a PR that is smoothly reviewed in a shorter time.

Issue/Pull Request Labels
~~~~~~~~~~~~~~~~~~~~~~~~~

Issues and PRs are labeled on GitHub so that they can be grouped, filtered and better maintained.
For instance, a label can indicate that a ticket needs response from the PR author, or that an issue needs immediate action in case of a critical bug.
Please refer to the `list of lables on GitHub <https://github.com/chainer/chainer/labels>`_.

> **Note that revision releases cannot include PRs labeled as a feature or compability breaking.**

Reviewer: Can't we remove this sentence? Recently we do not make revision releases, so almost all releases actually include feature addition (for compatibility breaks, it would be meaningful to mention that all minor releases cannot include them).

Author: Removed.

Issues and Pull Requests
------------------------

First steps in contributing to Chainer often involve filing an issue or creating a PR.
This section describes how to do so.

How to File an Issue
~~~~~~~~~~~~~~~~~~~~

To file an issue on GitHub, you often only need to follow instructions given by the template.
Write precise explanations on how you want Chainer to behave or include necessary and sufficient conditions to reproduce the bugs.
Feature requests should include **what** you want to do and preferably **why**.
You may additionally suggest **how**.

.. warning::

   If you have a question regarding the usage of Chainer, it is recommended to send a post to `StackOverflow <https://stackoverflow.com/>`_ or the `Chainer User Group <https://groups.google.com/forum/#!forum/chainer>`_ instead of the issue tracker.
   The issue tracker is not a place to share knowledge on practices.

How to Send a Pull Request
~~~~~~~~~~~~~~~~~~~~~~~~~~

If you can write code to fix an issue, it is encouraged to send a PR.

In that case, confirm the following points before starting to write any code.

> - Read the :ref:`coding-guide` and the :ref:`testing-guide`.

Reviewer: ```suggestion
- Read :ref:`coding-guide` and :ref:`testing-guide`.
```
Looking at the rendered document, I feel it's more natural as a hyperlinked document to show the linked section names without articles here.

.. note::

   Coding guidelines are updated in v3.0.

>    Those who have contributed to older versions may want to read the updated guidelines.

Reviewer: Can we remove this note because v3 is anyway very old?

Author: I agree, removed the note.

  $ autopep8 path/to/your/code.py
  $ flake8 path/to/your/code.py

``autopep8`` can automatically correct Python code to conform to the PEP 8 style guide::

  $ autopep8 --in-place path/to/your/code.py

> The ``flake8`` command lets you know which parts of your code that is not following the style guidelines.

Reviewer: Is this `which` (or `that`) needed?

Author: Fixed, and also `is` -> `are`.

This branch contains the most up-to-date source tree that includes features newly added after the latest major version.

The stable version is developed on the ``vN`` branch where "N" reflects the version number (*versioned branch*).
For example, v3.0.0, v3.1.0, and v3.2.0 are developed on the ``v3`` branch.

A PR from a contributor should in general be targeting the ``master`` branch.
If the change can and should be applied to the stable version in addition, a member from the core team will make sure it is backported to be included in the next revision update.

If the change is only applicable to the stable version and not to the ``master`` branch, please send it to the versioned branch.
We basically only accept changes to the latest versioned branch (where the stable version is developed) unless the fix is critical.

If you want to introduce a new feature in the ``master`` branch to the current stable version, please send a *backport PR* to the stable version (the latest ``vN`` branch).
See the next section for details.

*Note: a change that can be applied to both branches should be sent to the* ``master`` *branch.*

> *Each release of the stable version is also merged with the development version so that changes are reflected in the next major version.*

Reviewer: Is it true?

Author: Thanks, it seems to be old. Deleted this line.

Feature Backport Pull Requests

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: Shoud we mention the backport tool?

Author: Thanks for the tip. I added a sentence dedicated to it.

In general, new features in the development branch are not backported to the stable versions.
If such backports can be motivated and are necessary however, they are welcomed.

> In that a case, a backport PR must be sent to the latest ``vN`` branch.

Reviewer: ```suggestion
In such a case, a backport PR must be sent to the latest ``vN`` branch.
```

If you want to introduce a new feature in the ``master`` branch to the current stable version, please send a *backport PR* to the stable version (the latest ``vN`` branch).
See the next section for details.

*Note: a change that can be applied to both branches should be sent to the* ``master`` *branch.*

Feature Backport Pull Requests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In general, new features in the development branch are not backported to the stable versions.
If such backports can be motivated and are necessary however, they are welcomed.
In such a case, a backport PR must be sent to the latest ``vN`` branch.
**Note that we do not accept any feature backport PRs to older versions because we are not running quality assurance workflows (e.g. CI) for older versions which means that we cannot ensure that the PR is correctly ported.**

There are some rules on sending backport PRs.

* Prefix the PR title with **[backport]**.
* Include the original PR number in the PR description, e.g. "This is a backport of #XXXX".
* (Optional) Write in the PR description, the motivation behind the backport.

> There is a `backport tool https://github.com/chainer/backport`_ maintained by the core team that automates the process of creating backport PRs conforming to the rules above.

Reviewer: ```suggestion
There is a `backport tool <https://github.com/chainer/backport>`_ maintained by the core team that automates the process of creating backport PRs conforming to the rules above.
```
It looks the build fails at this line.

Author: Oh, thanks.

   Docstrings (documentation comments in the source code) are collected from the installed Chainer module.
   If you modified docstrings, make sure to install the module (e.g., using `pip install -e .`) before building the documentation.

.. note::

>    If you are unsure about how to write the documentation or failt to build it locally, you can submit a PR without documentations.

Reviewer: ```suggestion
   If you are unsure about how to write the documentation or failed to build it locally, you can submit a PR without documentation.
```

* Open-sourcing an external example
* Writing a post about Chainer

Development Cycle
-----------------

This section explains the development process of Chainer.

Versioning
~~~~~~~~~~

The versioning of Chainer follows `PEP 440 <https://www.python.org/dev/peps/pep-0440/>`_ and a part of `Semantic versioning <https://semver.org/>`_.
The version number consists of three or four parts: ``X.Y.Zw`` where ``X`` denotes the **major version**, ``Y`` denotes the **minor version**, ``Z`` denotes the **revision number**, and the optional ``w`` denotes the pre-release suffix.
While the major, minor, and revision numbers follow the rule of semantic versioning, the pre-release suffix follows PEP 440, the Python community standards.

**Note that a major update basically does not contain compatibility-breaking changes from the last release candidate (RC).**
This is not a strict rule, though; if there is a critical bug in the API that need to be fixed for the major version, breaking changes may be introduced.

For more on backward compatibility, please refer to the :ref:`compatibility`.

.. _contrib-release-cycle:

Release Cycle

> ~~~~~~~~~~~~~

Reviewer: We need to revisit this section; currently we are not following this 3-month cycle.

Author: How would you like to change it?

Reviewer: - stable versions are a series of minor updates (can be revision one, but not so usually)
- we may have more than one beta release (fix the table or think about another way to explain the parallel schedule)

Author: Thanks for the clarification. I made some changes as follows.

> - stable versions are a series of minor updates (can be revision one, but not so usually)

Fixed.

> - we may have more than one beta release (fix the table or think about another way to explain the parallel schedule)

Tried to explain it in the following body and wrote right before the table that this is an example and thus isn't always exact.

Whether these releases are revision or minor is determined based on the contents of each update.

Note that there are only three stable releases for the versions ``X.x.x``.
During the parallel development of ``Y.0.0`` and ``Z.0.0a1``, the version ``Y`` is treated as an **almost-stable version** and ``Z`` is treated as a development version.

If there is a critical bug found in ``X.x.x`` after stopping the development of version ``X``, we may release a hot-fix for this version at any time.

A `milestone for each upcoming release is published on GitHub <https://github.com/chainer/chainer/milestones>`_.
The GitHub milestones are used to group issues and PRs belonging to a release.

.. _contrib-git-branches:

Git Branches
~~~~~~~~~~~~

The ``master`` branch is used to develop pre-release versions.
It means that **alpha, beta, and RC updates are developed at the** ``master`` **branch**.
This branch contains the most up-to-date source tree that includes features newly added after the latest major version.

The stable version is developed on the ``vN`` branch where "N" reflects the version number (*versioned branch*).

> For example, v3.0.0, v3.0.1, and v3.0.2 are developed on the ``v3`` branch.

Reviewer: Better to use minor version numbers since they are more realistic for now.

Author: Fixed.

> A PR should from a contributor should in general be targeting the ``master`` branch.

Reviewer: ```suggestion
A PR from a contributor should in general be targeting the ``master`` branch.
```
